tfs/src/main/java/hudson/plugins/tfs/rm/ConnectReleaseWebHookEvent.java

```diff
@@ -0,0 +1,319 @@
+    }
+
+    private void deleteReleaseWebHook(final ReleaseWebHookResource resource) {
+        if (resource == null) {
+            throw new InvalidParameterException("event parameter is null");
+        }
+
+        final TeamPluginGlobalConfig config = TeamPluginGlobalConfig.get();
+        if (config == null) {
+            throw new InternalError("Cannot load TFS global configuration");
+        }
+
+        String webHookNameToDelete = resource.getWebHookName();
+        ReleaseWebHook webHook = validateAndGetReleaseWebHookByName(webHookNameToDelete);
+
+        for (final Item projectItem : Jenkins.getActiveInstance().getAllItems()) {
+            if (projectItem instanceof AbstractProject) {
+                AbstractProject project = (AbstractProject) projectItem;
+
+                ReleaseWebHookAction action = getReleaseWebHookActionFromProject(project);
+                if (action != null) {
+                    for (ReleaseWebHookName webHookName : action.getWebHookNames()) {
+                        if (webHookName.getWebHookName().equalsIgnoreCase(webHookNameToDelete)) {
+                            logger.fine(String.format("WebHook %s is referenced in project %s. Cannot delete until all the references are removed.", webHookNameToDelete, project.getName()));
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+
+        List<ReleaseWebHook> releaseWebHooks = ReleaseWebHookHelper.getReleaseWebHookConfigurations();
+        releaseWebHooks.remove(webHook);
+        ReleaseWebHookHelper.saveReleaseWebHookConfigurations(releaseWebHooks);
+    }
+
+    private JSONObject listReleaseWebHook() {
+        JSONArray webHooks = new JSONArray();
+
+        for (ReleaseWebHook webHookConfig : ReleaseWebHookHelper.getReleaseWebHookConfigurations()) {
+            JSONObject webHook = new JSONObject();
+            webHook.put("WebHookName", webHookConfig.getWebHookName());
+            webHook.put("PayloadUrl", webHookConfig.getPayloadUrl());
+
+            webHooks.add(webHook);
+        }
+
+        JSONObject result = new JSONObject();
+        result.put("ReleaseWebHooks", webHooks);
+
+        return result;
+    }
+
+    private void linkWebHook(final ReleaseWebHookResource resource) {
+        if (resource == null) {
+            throw new InvalidParameterException("event parameter is null");
+        }
+
+        AbstractProject project = validateAndGetJenkinsProject(resource);
+        String webHookNameToLink = resource.getWebHookName();
+
+        // validate the webhook exists in the server config
+        ReleaseWebHook webHook = validateAndGetReleaseWebHookByName(webHookNameToLink);
+
+        ReleaseWebHookAction action = getReleaseWebHookActionFromProject(project);
+
+        if (action != null && action.getWebHookNames() != null) {
+            for (ReleaseWebHookName webHookName : action.getWebHookNames()) {
```

Comment:
why do we need ReleaseWebHookName type? can't we use ReleaseWebHook instead?

Reply:
No. ReleaseWebHookName is part of ReleaseWebHookAction. ReleaseWebHookAction does not refer to actual releaseWebHook. It refer only to the name. Hence we need to the ReleaseWebHookName type.